modules/integration/tests-common/jacoco-report-generator/src/main/java/org/wso2/carbon/identity/jacoco/ReportGenerator.java

```diff
@@ -0,0 +1,198 @@
+        if (!reportDir.exists() && !reportDir.mkdirs()) {
+            throw new RuntimeException("Failed to create report directory: " + reportDir.getAbsolutePath());
+        }
+    }
+
+    /**
+     * Create the report.
+     *
+     * @throws IOException
+     */
+    public void create() throws IOException {
+
+        // Read the jacoco.exec file. Multiple data files could be merged
+        // at this point
+        loadExecutionData();
+
+        // Run the structure analyzer on a single class folder to build up
+        // the coverage model. The process would be similar if your classes
+        // were in a jar file. Typically you would create a bundle for each
+        // class folder and each jar you want in your report. If you have
+        // more than one bundle you will need to add a grouping node to your
+        // report
+        final IBundleCoverage bundleCoverage = analyzeStructure();
+
+        createReport(bundleCoverage);
+    }
+
+    private void createReport(final IBundleCoverage bundleCoverage) throws IOException {
+
+        // Create a concrete report visitor based on some supplied
+        // configuration. In this case we use the defaults
+        try (FileOutputStream fos = new FileOutputStream(xmlReport)) {
+            final XMLFormatter xmlFormatter = new XMLFormatter();
+            final IReportVisitor visitor = xmlFormatter.createVisitor(fos);
+
+            // Initialize the report with all the execution and session
+            // information. At this point the report doesn't know about the
+            // structure of the report being created
+            visitor.visitInfo(execFileLoader.getSessionInfoStore().getInfos(),
+                    execFileLoader.getExecutionDataStore().getContents());
+
+            // Populate the report structure with the bundle coverage information.
+            // Call visitGroup if you need groups in your report.
+            visitor.visitBundle(bundleCoverage,
+                    new DirectorySourceFileLocator(null, "utf-8", 4));
+
+            // Signal end of structure information to allow report to write all
+            // information out
+            visitor.visitEnd();
+        }
+    }
+
+    private void loadExecutionData() throws IOException {
+
+        execFileLoader = new ExecFileLoader();
+        execFileLoader.load(executionDataFile);
+    }
+
+    private IBundleCoverage analyzeStructure() throws IOException {
+
+        final CoverageBuilder coverageBuilder = new CoverageBuilder();
+        final Analyzer analyzer = new Analyzer(execFileLoader.getExecutionDataStore(), coverageBuilder);
+
+        List<File> jarFilesToAnalyze = new ArrayList<>();
+        List<File> classFilesToAnalyze = new ArrayList<>();
+
+        for (File classDirectory : classDirectories) {
+            // Jar files to analyze
+            File[] files = classDirectory.listFiles((dir, name) -> name.startsWith("org.wso2.carbon") && !name.contains(".stub_"));
```

Comment:
have we considered ignoring carbon ui components?

Reply:
We can ignore them during the report analysis level at the module repository level.